tests/unit/test_functools_utils.py

```diff
@@ -0,0 +1,50 @@
+import pytest
+
+from databuilder.functools_utils import singledispatchmethod_with_cache
+
+
+@pytest.fixture
+def TestClass():
+    COUNTER = 0
+
+    class TestClass:
+        @singledispatchmethod_with_cache
+        def test(self, value):
+            assert False
+
+        @test.register(str)
+        def test_str(self, value):
+            # Use a shared counter to give different results for each call
+            nonlocal COUNTER
+            COUNTER += 1
+            return value, COUNTER
+
+    return TestClass
+
+
+def test_results_are_cached(TestClass):
+    obj = TestClass()
+    assert obj.test("hello") == obj.test("hello")
```

Comment:
I'm not sure that would be testing what we care about here. You can imagine some variant on the cache function which pickles/unpickles its values when storing them. This would return distinct objects each time, but with identical values. So I think it's the values we care about.

The test method is designed (using the `COUNTER` trick) to return distinct values each time it's called so checking that the value is the same is sufficient to determine that the method hasn't been called a second time.

I might be misunderstanding your suggestion though?

Reply:
No, you're not misunderstanding! I didn't understand the context well enough, and so thought that the identity of the returned values mattered when creating CTEs.

Reply:
Ah no, that's a good point! I was thinking in terms of the generic `functools.cache` behaviour. But you're quite right that for some kinds of SQLAlchemy object identity matters as well.

So actually I think you're right: we should switch all these out for `is/is not`. Thanks!